README.md
# stratachalup.github.io
 


cgi-bin/ChangeLog
Version 0.42
------------

1. renamed cgi-bin ("wiki")
2. added /usr/local/bin to the PATH in PageArchiveRCS|CVS.pm
3. bugfix AtisWiki "sub asLink": remove "::" in the variables
   $self, $pa (reported by Jonas Karlson)

Version 0.41alpha
-----------------

1. Fixed a Bug in Wiki.pm (reported by Jonas Karlson)

Version 0.4alpha
----------------

1. moved addBacklinks to PageArchive.pm
2. Main Wiki page defined via $MainPage, not the
   name of the script. 


Version 0.3alpha
----------------

1. changed "Backlinks" to do simple searching until I find the
   Backlinks-BUG.   


Version 0.2alpha
----------------

1. added Umlauts to FormatWiki.html

2. added Umlauts to $LinkPattern

3. use application/x-www-form-urlencoded for posting instead
   of multipart/form-data (Problems with some Versions if Lynx, 
   reported by Andreas Gross)

4. case-insensitive searching (FIX by Simon Michael)

5. added file: URLs (FIX by Simon Michael)

6. FormatRawHtml: send HTTP-Header before sending the HTML-Data
   (Bugfix by  Larry Kollar <Larry.Kollar@arris-i.com>)

7. FormatWiki: bullet list: <TAB> optional (problems with some Versions
   of Lynx, reported by Andreas Gross)

8. $LinkPattern: allows CAPSWord, but not CAPSWORD (Simon Michael)

9. Moved printHeaderText, printFooterText to "LookAndFeel.pm"

10. created LookAndFeelJoyful.pm, LookAndFeelRO.pm
    ("Joyful" Edit-box by Simon Michael)

11. Markup: Lines starting with "|" : no Parsing for WikiLinks 
    (requested by Andreas Gross)

12. Lines can be joined with "\" at the end of the line.
    BugFIX.

13. Quoting of Wikinames: $WikiName will not be konverted to a Link
    (requested by Andreas Gross)

14. copied locking from CVWiki



cgi-bin/Wiki.pm
# The Wiki Perl Module Library: Wiki.pm  
#   Copyright (C) 1998 Marcus Denker
#   Copyright (C) 1998 ATIS Department of Computer Science, 
#                      University of Karlsruhe / Germany  
#   based on 
#   ->the LGPLed CVWiki CVS-patches (C) 1997 Peter Merel
#   ->and The Original WikiWikiWeb  (C) Ward Cunningham
#   (code reused with permission)
#
#   Written by:  Marcus Denker <marcus@ira.uka.de>
#   Created: Aug 1998
#
#   This library is free software; you can redistribute it and/or
#   modify it under the terms of the GNU Library General Public
#   License as published by the Free Software Foundation; either
#   version 2 of the License, or (at your option) any later version.
#   
#   This library is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   Library General Public License for more details.
#
#   You should have received a copy of the GNU Library General Public
#   License along with this library; if not, write to the Free
#   Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
package   Wiki;
require   Exporter;
@ISA    = qw(Exporter);
@EXPORT = qw(&Print &ToHTML &calcDay &quoteHtml &r_u &addToRecentChanges
	    &renamePage &RequestLock &ReleaseLock);

use strict;

# fallback Print functions, called when no TypeInfo available.
sub Print
  {
    my $id = shift;
    &printHeaderText($id);
    print "Format-Error: Wiki tried to show a Page without TypeInformation!<br><hr>";
    &printFooterText($id);
  }
    
sub ToHTML
  {
    return "Format-Error: Wiki tried to show Body of a Page without TypeInformation!<br>";
  }


sub calcDay 
  {
    my ($sec, $min, $hour, $mday, $mon, $year) = localtime($^T);
    return ("January", "February", "March", "April", "May", "June", 
                 "July", "August", "September", "October", "November", 
                 "December")[$mon]. " " . $mday . ", " . ($year+1900);
  }



sub quoteHtml
  {
    $_ = shift;
    s/&/&amp;/g;
    s/</&lt;/g;
    s/>/&gt;/g;
    return $_;
  }

sub r_u 
  {
    $_ = shift;
    s/_/ /g;
    s/^ //g; 
    s/ $//g;
    return $_;
  }


sub addToRecentChanges
  {
    my $id = shift;
    my $date = &calcDay;
    my %rc = $::pa->getPage("RecentChanges");
    
    $rc{text} =~ s/\t\* $id .*\n//g;
    $rc{text} .= "\n$date\n\n" unless $rc{text} =~ /$date/;
    $rc{text} .= "\t* $id . . . . . . $::page{logname}\n";
    

    $rc{type} = "wiki";
    $::pa->setPage("RecentChanges","RecentChanges",%rc);
  }


sub renamePage 
  {
    my ($old, $id) = @_;
    # Search for the old Name in all Pages
    my @pages = $::pa->search("$old");
    foreach (@pages) {
      my %pg = $::pa->getPage($_);
      if ($pg{text} =~ $old) { 
	$pg{text} =~ s/$old/$id/geo }
      $::pa->setPage($_, "Rename", %pg);
    }
    my %oldPage = $::pa->getPage($old);
    $::pa->removePage($old);
    &addToRecentChanges($id);
    $::pa->addBacklinks($id, $oldPage{text});
    $::pa->setPage($id, "Wiki Edit", %oldPage);

    my @links = ($oldPage{text} =~ /\b($::LinkPattern)\b/gs);
    foreach (@pages) {
      my %pg = $::pa->getPage($_);
      $pg{backlinks} =~ s/$old/$id/gs; #FIXME! 
      $::pa->setPage($_, "Rename", %pg);
    }
  
 }


 
sub RequestLock 
{

  my ($n) = 0;
  while (mkdir($::LockDirectory, 0555) == 0) 
    {
      #
      # EEXIST == 17 is OK, try later.
      #
      $! == 17 || die("can't make $::LockDirectory: $!\n");
      $n++ < 30 || die("timed out waiting for $::LockDirectory\n");
      sleep(1);
    }
}


sub ReleaseLock {
  rmdir($::LockDirectory);
}

1;



cgi-bin/FormatAscii.pm
# The Wiki Perl Module Library: FormatAscii.pm -- ASCII Markup
#   Copyright (C) 1998 Marcus Denker
#   Copyright (C) 1998 ATIS Department of Computer Science, 
#                      University of Karlsruhe / Germany  
#   
#   Written by:  Marcus Denker <marcus@ira.uka.de>
#   Created: Aug 1998
#
#   This file is part of AtisWiki 
#
#   This library is free software; you can redistribute it and/or
#   modify it under the terms of the GNU Library General Public
#   License as published by the Free Software Foundation; either
#   version 2 of the License, or (at your option) any later version.
#   
#   This library is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   Library General Public License for more details.
#
#   You should have received a copy of the GNU Library General Public
#   License along with this library; if not, write to the Free
#   Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
package   FormatAscii;
require   Exporter;
@ISA    = qw(Exporter);
@EXPORT = qw (&AsciiToHTML &AsciiPrint);

use strict;
use Wiki;
use LookAndFeel;

#---------
#
# Format: Ascii
#
# show data as ASCII-text.
#

$::knownTypes{Ascii}=1;

sub AsciiToHTML
  {
    my $string = &quoteHtml(shift);
    return "<CODE><PRE> $string </PRE></CODE>";  
  }

sub AsciiPrint
  {
    my $id = shift;
    &printHeaderText($id);
    print AsciiToHTML($::page{text}),"<hr>";
    &printFooterText($id);
  }
1;


cgi-bin/FormatRawhtml.pm
# The Wiki Perl Module Library: FormatRawhtml.pm
#   Copyright (C) 1998 Marcus Denker
#   Copyright (C) 1998 ATIS Department of Computer Science, 
#                      University of Karlsruhe / Germany  
#   
#   Written by:  Marcus Denker <marcus@ira.uka.de>
#   Created: Aug 1998
#
#   This file is part of AtisWiki 
#
#   This library is free software; you can redistribute it and/or
#   modify it under the terms of the GNU Library General Public
#   License as published by the Free Software Foundation; either
#   version 2 of the License, or (at your option) any later version.
#   
#   This library is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   Library General Public License for more details.
#
#   You should have received a copy of the GNU Library General Public
#   License along with this library; if not, write to the Free
#   Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
package   FormatRawhtml;
require   Exporter;
@ISA    = qw(Exporter);
@EXPORT = qw(&rawhtmlToHTML &rawhtmlPrint);

use strict;
use Wiki;

#----
#
# Format: rawhtml
#
# show data as is, no header, no footer. 
# use this as a template for your own formats!

$::knownTypes{rawhtml}=1;
sub rawhtmlToHTML
  {
    my $string = shift;
    return $string;
  }

sub rawhtmlPrint
  {
    print $::q->header();
    print $::page{text};
  }

1;


cgi-bin/LookAndFeelJoyful.pm
# The Wiki Perl Module Library  
#   Copyright (C) 1999 Simon Michael
#   Copyright (C) 1998 Marcus Denker 
#   Copyright (C) 1998 ATIS Department of Computer Science, 
#                      University of Karlsruhe / Germany  
#   Copyright (C) 1997 Peter Merel
#   Copyright (C) 1996 Ward Cunningham
#   
#   Written by:  Marcus Denker <marcus@ira.uka.de>
#   Created: Aug 1998
#
#   based on 
#   ->the LGPLed CVWiki CVS-patches (C) 1997 Peter Merel
#   ->and The Original WikiWikiWeb  (C) Ward Cunningham
#   (code reused with permission)
#
#   This file is part of AtisWiki 
#
#   This library is free software; you can redistribute it and/or
#   modify it under the terms of the GNU Library General Public
#   License as published by the Free Software Foundation; either
#   version 2 of the License, or (at your option) any later version.
#   
#   This library is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   Library General Public License for more details.
#
#   You should have received a copy of the GNU Library General Public
#   License along with this library; if not, write to the Free
#   Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
package   LookAndFeel;
require   Exporter;
@ISA    = qw(Exporter);
@EXPORT = qw (&printHeaderText &printFooterText $LogoImage);

use strict;
use Wiki;

#
# LookAndFeelJoyful 
#


my $LogoUrl = '/uw_sm.gif';
my $LogoImage = "<img src=\"$LogoUrl\" alt=\"Home\" border=0>";

$::editallowed=1;

sub printHeaderText 
  {
    my $id = shift;
    my $title = r_u($id);
    print $::q->header(),
         $::q->start_html('-title'   => "$::self: $title", 
                        '-BGCOLOR' => 'white');
    print '<table width="100%" border="0" cellspacing="0" cellpadding="0"><tr>', "\n";
    print "<td>",
         $::q->h1("<a href=\"$::self\" target=\"\_top\">$LogoImage</a>",
	        "<a href=\"$::self\?action=Backlinks&id=$id\" target=\"\_top\">$title</a>"),
                "</td>\n<td align=\"right\">",
         $::q->startform(-target=>"_top"),
         "Search:",
         $::q->textfield(-name=>'search',
                       -size=>20,
                       -target=>"_top"),
         $::q->endform,
         "</td></tr></table><hr>\n";
  }

   
sub printFooterText 
  {
    my $id = shift;
    my $title = r_u($id);
    print '<table width="100%" border="0" cellspacing="0" cellpadding="2">', "\n";
    print '<tr><td bgcolor="ccffaa">', "\n";
    print $::q->startform("POST","$::self",
                        "multipart/form-data");
    $::q->param("title","$id");
    print $::q->hidden("title") , "\n";
    print "<strong><a href=\"$::self\?action=Edit&id=$id\" target=\"_top\">Edit<\/a></strong> this page:";
    print "&nbsp;\|&nbsp;<strong><a href=\"$::self\?action=Rename&id=$id\">Rename<\/a></strong> this page";
    if ($::pa->doBackup) {
      print "&nbsp;\|&nbsp;<strong><a href=\"$::self\?action=Backup&id=$id\">Backup Copies<\/a></strong>";
    }
    print "&nbsp;\|&nbsp;<strong><a href=\"$::self\?WikiHelp\">Help<\/a></strong> for this site, wikis, editing, etc.";
    print "<br>\n";
    print $::q->textarea('text', $::page{text}, 4, 80);
    print "\n<br><big>",
          $::q->submit(-name=>'Save'),
          "&nbsp;",
          $::q->reset, "</big>";

    if (keys %::knownTypes > 1)  {
      $::page{type} ||=  $::defaultType; 
      print "&nbsp;Page type: ",
          $::q->popup_menu('type_menu',[keys %::knownTypes],"$::page{type}");
    } 
    print "&nbsp;<em>Last edited $::page{date}</em>";

    print $::q->endform;

    print "\n", '</td></tr></table>';
    print $::q->end_html;
}

1;


cgi-bin/LookAndFeelNormal.pm
# The Wiki Perl Module Library  
#   Copyright (C) 1998 Marcus Denker
#   Copyright (C) 1998 ATIS Department of Computer Science, 
#                      University of Karlsruhe / Germany  
#   Copyright (C) 1997 Peter Merel
#   Copyright (C) 1996 Ward Cunningham
#   
#   Written by:  Marcus Denker <marcus@ira.uka.de>
#   Created: Aug 1998
#
#   based on 
#   ->the LGPLed CVWiki CVS-patches (C) 1997 Peter Merel
#   ->and The Original WikiWikiWeb  (C) Ward Cunningham
#   (code reused with permission)
#
#   This file is part of AtisWiki 
#
#   This library is free software; you can redistribute it and/or
#   modify it under the terms of the GNU Library General Public
#   License as published by the Free Software Foundation; either
#   version 2 of the License, or (at your option) any later version.
#   
#   This library is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   Library General Public License for more details.
#
#   You should have received a copy of the GNU Library General Public
#   License along with this library; if not, write to the Free
#   Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
package   LookAndFeel;
require   Exporter;
@ISA    = qw(Exporter);
@EXPORT = qw (&printHeaderText &printFooterText $LogoImage);

use strict;
use Wiki;

#
# LookAndFeel: normal. 
#


my $LogoUrl = '/uw_sm.gif';
my $LogoImage = "<img src=\"$LogoUrl\" alt=\"Home\" border=0 align=\"right\">";
$::editallowed=1;

sub printHeaderText
  {
    my $id = shift;
    my $title = r_u($id);
    print $::q->header(),  
          $::q->start_html('-title'   => "$::self: $title",
                         '-BGCOLOR' => 'white'),
          $::q->h1("<a href=\"$::self\" target=\"\_top\">$LogoImage</a>
    <a href=\"$::self\?action=Backlinks&id=$id\" target=\"\_top\">$title</a>");
  }
   
sub printFooterText
  {
    my $id = shift;
    print $::q->startform(-target=>"_top"),
          "<a href=\"$::self\?action=Edit&id=$id\" target=\"_top\">EditText<\/a> ",
          "of this Page (last edited $::page{date})<br>",
          "<a href=\"http://www.opencontent.org/\"> <img
          src=\"http://www.opencontent.org/takeone.gif\" border=0 align=\"right\"
          alt=\"Take One!\"></a>",
          "Search:",
          $::q->textfield(-name=>'search',
                        -size=>20,
                        -target=>"_top"),
          $::q->endform,
          $::q->end_html;
  }

1;


cgi-bin/LookAndFeel.pm
# The Wiki Perl Module Library  
#   Copyright (C) 1999 Simon Michael
#   Copyright (C) 1998 Marcus Denker 
#   Copyright (C) 1998 ATIS Department of Computer Science, 
#                      University of Karlsruhe / Germany  
#   Copyright (C) 1997 Peter Merel
#   Copyright (C) 1996 Ward Cunningham
#   
#   Written by:  Marcus Denker <marcus@ira.uka.de>
#   Created: Aug 1998
#
#   based on 
#   ->the LGPLed CVWiki CVS-patches (C) 1997 Peter Merel
#   ->and The Original WikiWikiWeb  (C) Ward Cunningham
#   (code reused with permission)
#
#   This file is part of AtisWiki 
#
#   This library is free software; you can redistribute it and/or
#   modify it under the terms of the GNU Library General Public
#   License as published by the Free Software Foundation; either
#   version 2 of the License, or (at your option) any later version.
#   
#   This library is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   Library General Public License for more details.
#
#   You should have received a copy of the GNU Library General Public
#   License along with this library; if not, write to the Free
#   Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
package   LookAndFeel;
require   Exporter;
@ISA    = qw(Exporter);
@EXPORT = qw (&printHeaderText &printFooterText $LogoImage);

use strict;
use Wiki;

#
# LookAndFeelJoyful 
#


my $LogoUrl = 'http://kgate.virtual.net/vctiny.gif';
my $LogoImage = "<img src=\"$LogoUrl\" alt=\"Home\" border=0>";

$::editallowed=1;

sub printHeaderText 
  {
    my $id = shift;
    my $title = r_u($id);
    print $::q->header(),
         $::q->start_html('-title'   => "$::self: $title", 
                        '-BGCOLOR' => 'white');
    print '<table width="100%" border="0" cellspacing="0" cellpadding="0"><tr>', "\n";
    print "<td>",
         $::q->h1("<a href=\"$::self\" target=\"\_top\">$LogoImage</a>",
	        "<a href=\"$::self\?action=Backlinks&id=$id\" target=\"\_top\">$title</a>"),
                "</td>\n<td align=\"right\">",
         $::q->startform(-target=>"_top"),
         "Search:",
         $::q->textfield(-name=>'search',
                       -size=>20,
                       -target=>"_top"),
         $::q->endform,
         "</td></tr></table><hr>\n";
  }

   
sub printFooterText 
  {
    my $id = shift;
    my $title = r_u($id);
    print '<table width="100%" border="0" cellspacing="0" cellpadding="2">', "\n";
    print '<tr><td bgcolor="ccffaa">', "\n";
    print $::q->startform("POST","$::self",
                        "multipart/form-data");
    $::q->param("title","$id");
    print $::q->hidden("title") , "\n";
    print "<strong><a href=\"$::self\?action=Edit&id=$id\" target=\"_top\">Edit<\/a></strong> this page:";
    print "&nbsp;\|&nbsp;<strong><a href=\"$::self\?action=Rename&id=$id\">Rename<\/a></strong> this page";
    if ($::pa->doBackup) {
      print "&nbsp;\|&nbsp;<strong><a href=\"$::self\?action=Backup&id=$id\">Backup Copies<\/a></strong>";
    }
    print "&nbsp;\|&nbsp;<strong><a href=\"$::self\?WikiHelp\">Help<\/a></strong> for this site, wikis, editing, etc.";
    print "<br>\n";
    print $::q->textarea('text', $::page{text}, 4, 80);
    print "\n<br><big>",
          $::q->submit(-name=>'Save'),
          "&nbsp;",
          $::q->reset, "</big>";

    if (keys %::knownTypes > 1)  {
      $::page{type} ||=  $::defaultType; 
      print "&nbsp;Page type: ",
          $::q->popup_menu('type_menu',[keys %::knownTypes],"$::page{type}");
    } 
    print "&nbsp;<em>Last edited $::page{date}</em>";

    print $::q->endform;

    print "\n", '</td></tr></table>';
    print $::q->end_html;
}

1;
